django/db/migrations/optimizer.py

```diff
@@ -44,20 +44,29 @@ def optimize_inner(self, operations, app_label=None):
         """
         new_operations = []
         for i, operation in enumerate(operations):
+            right = True  # Should we reduce on the right or on the left.
             # Compare it to each operation after it
             for j, other in enumerate(operations[i + 1:]):
                 in_between = operations[i + 1:i + j + 1]
-                result = operation.reduce(other, in_between, app_label)
+                result = operation.reduce(other, app_label)
                 if isinstance(result, list):
-                    # Optimize! Add result, then remaining others, then return
-                    new_operations.extend(result)
-                    new_operations.extend(in_between)
+                    if right:
+                        new_operations.extend(in_between)
+                        new_operations.extend(result)
+                    elif all(op.reduce(other, app_label) is True for op in in_between):
```

Comment:
I believe this was meant to be a `references` check instead of a `reduce` check? Since `reduce` returns operations, not a boolean.

Something like "`not op.references(other)`" so that we can ensure that `other` can properly be pulled forwards